### PR TITLE
Special case slot 0 for block connectness check in ledger tool

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

Fix https://github.com/solana-labs/solana/issues/29799

#### Summary of Changes

Special case slot 0 for block connectness check.

Fixes #

### CODE REVIEW

**Comment by mvines** [comment]
Probably want to remove "[filter stake by min delegation for rewards](https://github.com/solana-labs/solana/pull/29860/commits/243000fa2015be652c7e6261875a1e6b4d1bd072)" from here!

File: ledger-tool/src/main.rs [hunk 1]
```diff
@@ -1066,16 +1066,15 @@ fn load_bank_forks(
     };
 
     if let Some(halt_slot) = process_options.halt_at_slot {
```
Comment (apfitzge) [inline ledger-tool/src/main.rs:1068]: Ideally we'd just check halt_slot != 0 here so we can make a comment about the special case, but I don't think rust allows chained conditionals with if let yet. An alternative is something like:
```
match process_options.halt_at_slot {
    // Sentinel value of `Some(0)` lets us skip additional checks and will not replay any transactions.
    Some(halt_slot) if halt_slot != 0 => { /*...*/ },
    _ => {},
}
```

Comment (apfitzge) [reply ledger-tool/src/main.rs:1068]: Also happy if we just add a comment above 1068 with your current changes.

Comment (HaoranYi) [reply ledger-tool/src/main.rs]: yeah. good idea.

File: ledger-tool/src/main.rs [hunk 2]
```diff
@@ -1065,18 +1065,25 @@ fn load_bank_forks(
         })
     };
 
-    if let Some(halt_slot) = process_options.halt_at_slot {
-        if halt_slot < starting_slot && halt_slot != 0 {
-            eprintln!(
+    match process_options.halt_at_slot {
+        // For Sentinel value of `Some(0)` and None of halt_at_slot, lets us skip
```
Comment (apfitzge) [inline ledger-tool/src/main.rs:1069]: Apologies for being nit-picky, but `Some(0)` and `None` are actually different here. In both we skip these checks. But `None` will replay all available transactions (effectively it's `Some(Slot::MAX)` whereas `Some(0)` replays nothing.

Comment (apfitzge) [reply ledger-tool/src/main.rs:1069]: code is fine, and much better than what I suggested. But just nit-picking the comment for clarity.

Comment (HaoranYi) [reply ledger-tool/src/main.rs]: yeah.

File: ledger-tool/src/main.rs [hunk 3]
```diff
@@ -1065,19 +1065,27 @@ fn load_bank_forks(
...
     };
 
-    if let Some(halt_slot) = process_options.halt_at_slot {
-        if halt_slot < starting_slot {
-            eprintln!(
+    match process_options.halt_at_slot {
+        // There are no needs to do the following checks for Sentinel value of `Some(0)` and None
+        // of halt_at_slot, because for Sentinel value of `Some(0)`, no transaction will be
+        // replayed, and for `None`, which implies `Some(Slot::MAX), all transactions that are
+        // available will be replayed.
```
Comment (steviez) [inline ledger-tool/src/main.rs:1072]:
```suggestion
        // Skip the following checks for sentinel values of Some(0) and None.
        // For Some(0), no slots will be be replayed after starting_slot.
        // For None, all available children of starting_slot will be replayed.
```
I like the general point of the comment; just tweaked the wording a little which IMO makes it a little more clear. Let me know what you both think

File: ledger-tool/src/main.rs [hunk 4]
```diff
@@ -1065,19 +1065,27 @@ fn load_bank_forks(
...
-            exit(1);
-        }
-        // Check if we have the slot data necessary to replay from starting_slot to <= halt_slot.
-        //  - This will not catch the case when loading from genesis without a full slot 0.
-        if !blockstore.slot_range_connected(starting_slot, halt_slot) {
-            eprintln!("Unable to load bank forks at slot {halt_slot} due to disconnected blocks.",);
-            exit(1);
+                exit(1);
+            }
+            // Check if we have the slot data necessary to replay from starting_slot to <= halt_slot.
```
Comment (steviez) [inline ledger-tool/src/main.rs:1082]:
```suggestion
            // Check if we have the slot data necessary to replay from starting_slot to >= halt_slot.
```
Just a swap on the sign; we stop once have hit or surpassed halt_slot

**Review by steviez — COMMENTED** [review]
Just a couple tweaks on the comments. Agreed with Andrew on the code; my mind didn't make the jump to do match here and it is cleaner than the nested `if`'s I would have coded up

File: ledger-tool/src/main.rs (hunk 3 above)
Comment (HaoranYi) [reply ledger-tool/src/main.rs]: yeah. I like your wording.

**Review by apfitzge — APPROVED** [review]
lgtm

**Review by steviez — APPROVED** [review]

**Pull request merged**